MetaMask/metamask-extension
# Bump `@metamask/transaction-controller` to `^23.0.0`

## **Description**

Currently, Extension can only process transactions for the active / globally selected network. This makes it not possible to facilitate multichain interactions without constantly rotating the active network or some other workaround that would hurt user experience. We want to keep Metamask competitive by making our multichain experience as seamless and frictionless as possible. This PR introduces an updated version of the TransactionController that is capable of handling transactions across multiple network clients concurrently which lays the groundwork for future multichain changes.

* Bumps `@metamask/transaction-controller` to `^23.0.0`
* Adds `TRANSACTION_MULTICHAIN` environment variable
  * when set, enables the TransactionController to submit, process, and track transactions concurrently on mutiple networks
  * defaults off

NOTE: This PR does not introduce a CI pipeline for building and testing with TRANSACTION_MULTICHAIN enabled because the SwapController still needs to be updated to be compatible. This feature flag defaults off and does not change the current user facing TransactionController behavior, so current testing pipelines are sufficient.

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/1990

## **Manual testing steps**

### Testing the single globally selected transaction flow
There should be no user facing difference in behavior. 

### Testing the multichain transaction flow
1. Make a build with `TRANSACTION_MULTICHAIN` set to 1
2. Switch to local testnet network
3. Send a transaction
4. Quickly, before the transaction is marked confirmed, switch the globally selected network to a network on a different chain
5. Wait a few moments
6. Turn off the local testnet so that Metamask cannot connect to it anymore
7. Switch back to the testnet network
8. The transaction should be confirmed

This demonstrates that the TransactionController was polling the local testnet network for transaction updates while it was not the active / globally selected network and was able to confirm the transaction onchain (as opposed to it confirming the transaction right as it became the active / globally selected network again since Metamask wouldn't be able to reach the local testnet anymore)

## **Screenshots/Recordings**

### **Before** 

https://github.com/MetaMask/metamask-extension/assets/918701/d4e72e76-e975-460f-a7c4-c294bea88708

Notice that the transaction was not confirmed when we switched to mainnet and that when we switched back to localhost that it confirmed instantly when we turned ganache back on, implying that the TransactionController is only polling for the currently active network.


### **After** (with TRANSACTION_MULTICHAIN flag enabled)

https://github.com/MetaMask/metamask-extension/assets/918701/46edc340-0c91-45cf-928f-20c185aa407e

Notice that the local testnet transaction was not confirmed before we switched to mainnet, but was confirmed by the time we switched back to it despite us being on mainnet and turning off ganache, implying that the TransactionController was polling for the local testnet transaction while mainnet was the globally selected network.

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
